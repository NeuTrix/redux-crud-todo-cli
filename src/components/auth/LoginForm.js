import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
// import functions
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import { withStyles } from '@material-ui/core/styles';
import FormControl from '@material-ui/core/FormControl';

import ProgressBar from '../buttons/ProgressBar';
import validateInput from '../../helpers/loginValidator';
import { colors } from '../../helpers/cssConstants';

const propTypes = {
	addFlashMessage: PropTypes.func.isRequired,
	authApi: PropTypes.instanceOf(Object).isRequired,
	classes: PropTypes.instanceOf(Object).isRequired,
	userLoginRequest: PropTypes.func.isRequired,
};

const contextTypes = {
	router: PropTypes.instanceOf(Object).isRequired,
};

class Loginform extends Component {
	constructor(props, context) {
		super(props, context);
		const { authApi } = this.props;
		this.state = {
			errors: { },
			identifier: '',
			isLoading: authApi.loginIsPosting,
			password: '',
		};
		this.onChange = this.onChange.bind(this);
		this.onSubmit = this.onSubmit.bind(this);
	}

	onChange(e) {
		e.preventDefault();
		this.setState({ [e.target.name]: e.target.value });
	}

	onSubmit(e) {
		const { router } = this.context;
		const { addFlashMessage, userLoginRequest } = this.props;

		e.preventDefault();

		if (this.isValid()) {
			// reset state if no errors
			this.setState({ errors: { }, isLoading: true });
			// pass the state forward for a login ...
			userLoginRequest(this.state)
				.then((res) => {
					addFlashMessage({
						text: `Hi ${res.data.username}! You're Logged In.`,
						type: 'success',
					});
					router.history.push('/todos');
					return res;
				})
				.catch((err) => {
					this.setState({ errors: err, isLoading: false });
					addFlashMessage({
						text: 'Invalid username, id or password. Try again.',
						type: 'error',
					});
				});
		}
	}

	isValid() {
		// collect the return objects from #validateInput
		const { errors, isValid } = validateInput(this.state);
		if (!isValid) {
			this.setState({ errors });
		}
		return isValid; // a boolean value from #validateInput
	}

	render() {
		const { classes } = this.props;
		const { errors, identifier, password, isLoading } = this.state;

		return (
			<FormControl
				className={`{loginGrid ${classes.grid}`}
				component="form"
				onSubmit={this.onSubmit}
			>
				<div
					className={`{loginTitle ${classes.loginTitle}`}
					variant="h2"
					color="secondary"
				>
					{ !isLoading
						? (
							<Typography variant="h4" color="secondary">
								{'Login'}
							</Typography>
						)
						: <ProgressBar />
					}
				</div>

				<TextField
					className={`{email ${classes.email}`}
					errors={errors.identifier}
					label="Username | Email"
					margin="normal"
					name="identifier"
					placeholder="enter username or email"
					required
					value={identifier}
					variant="outlined"
					onChange={this.onChange}
				/>

				<TextField
					className={`{password ${classes.password}`}
					errors={errors.password}
					label="Password"
					margin="normal"
					name="password"
					placeholder="enter password"
					required
					type="password"
					value={password}
					variant="outlined"
					onChange={this.onChange}
				/>
				<Link
					className={classes.regLink}
					to="/register"
				>
					<Typography>
						{'Click here for a new account'}
					</Typography>
				</Link>
					
				<Button
					className={`{submit ${classes.button}`}
					color="primary"
					component="button"
					disabled={isLoading}
					name="Log in"
					size="small"
					type="submit"
					variant="contained"
				>
					{'Go'}
				</Button>
			</FormControl>
		);
	}
}

const styles = theme => ({
	button: {
		color: '#fafafa',
		gridArea: 'button',
		marginTop: theme.spacing.unit * 2,
		placeSelf: 'center',
		width: 100,
	},
	email: { gridArea: 'email' },
	grid: {
		border: `1px solid ${colors._mintgreen}`,
		borderRadius: 5,
		display: 'grid',
		gridRowGap: theme.spacing.unit * 4,
		gridTemplateAreas: ` 
			'loginTitle'
			'email'
			'password'
			'button'
			'regLink'
		`,
		padding: 20,
		width: 300,
		[theme.breakpoints.up('lg')]: {
			width: 500,
		},
	},
	input: { display: 'none' },
	loginTitle: { gridArea: 'loginTitle' },
	password: { gridArea: 'password' },
	regLink: {
		gridArea: 'regLink',
		paddingTop: 25,
		placeSelf: 'center',
	},
});

Loginform.propTypes = propTypes;
Loginform.contextTypes = contextTypes;

export default withStyles(styles)(Loginform);
