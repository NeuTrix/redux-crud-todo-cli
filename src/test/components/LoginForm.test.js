import React from 'react';
import { render } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import LoginForm from '../../components/auth/LoginForm';

describe('The LoginForm', () => {
  const authApi = {
    loginIsPosting: true,
  };

  const addFlashMessage = () => f;
  const wrapper = render(
    <Router>
      <LoginForm 
        authApi={authApi}
        addFlashMessage={addFlashMessage}
      />
    </Router>
  );

  it('...renders the login form', () => {
    expect(wrapper).toMatchSnapshot();
  });

});