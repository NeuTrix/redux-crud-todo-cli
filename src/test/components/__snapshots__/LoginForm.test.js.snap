// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The LoginForm ...renders the login form 1`] = `
<div
  class="{loginGrid Loginform-grid-3"
>
  <div
    class="{loginTitle Loginform-loginTitle-5"
    color="secondary"
    variant="h2"
  >
    <div
      class="ProgressBar-root-10"
    >
      <div
        aria-valuenow="0"
        class="MuiLinearProgress-root-11 MuiLinearProgress-colorPrimary-12 ProgressBar-bar-9"
        role="progressbar"
      >
        <div
          class="MuiLinearProgress-bar-19 MuiLinearProgress-barColorPrimary-20 MuiLinearProgress-bar1Determinate-23"
          style="transform:scaleX(0)"
        />
      </div>
    </div>
  </div>
  <div
    class="MuiFormControl-root-28 MuiFormControl-marginNormal-29 {email Loginform-email-2"
  >
    <label
      class="MuiFormLabel-root-39 MuiFormLabel-required-44 MuiInputLabel-root-32 MuiInputLabel-formControl-33 MuiInputLabel-animated-36 MuiInputLabel-outlined-38"
      data-shrink="false"
    >
      Username | Email
      <span
        class="MuiFormLabel-asterisk-45"
      >
         *
      </span>
    </label>
    <div
      class="MuiInputBase-root-59 MuiOutlinedInput-root-46 MuiInputBase-formControl-60"
    >
      <fieldset
        aria-hidden="true"
        class="MuiPrivateNotchedOutline-root-76 MuiOutlinedInput-notchedOutline-53"
        style="padding-left:8px"
      >
        <legend
          class="MuiPrivateNotchedOutline-legend-77"
          style="width:0.01px"
        />
      </fieldset>
      <input
        aria-invalid="false"
        class="MuiInputBase-input-69 MuiOutlinedInput-input-54"
        name="identifier"
        placeholder="enter username or email"
        required=""
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="MuiFormControl-root-28 MuiFormControl-marginNormal-29 {password Loginform-password-6"
  >
    <label
      class="MuiFormLabel-root-39 MuiFormLabel-required-44 MuiInputLabel-root-32 MuiInputLabel-formControl-33 MuiInputLabel-animated-36 MuiInputLabel-outlined-38"
      data-shrink="false"
    >
      Password
      <span
        class="MuiFormLabel-asterisk-45"
      >
         *
      </span>
    </label>
    <div
      class="MuiInputBase-root-59 MuiOutlinedInput-root-46 MuiInputBase-formControl-60"
    >
      <fieldset
        aria-hidden="true"
        class="MuiPrivateNotchedOutline-root-76 MuiOutlinedInput-notchedOutline-53"
        style="padding-left:8px"
      >
        <legend
          class="MuiPrivateNotchedOutline-legend-77"
          style="width:0.01px"
        />
      </fieldset>
      <input
        aria-invalid="false"
        class="MuiInputBase-input-69 MuiOutlinedInput-input-54 MuiInputBase-inputType-72"
        name="password"
        placeholder="enter password"
        required=""
        type="password"
        value=""
      />
    </div>
  </div>
  <a
    class="Loginform-regLink-7"
    href="/register"
  >
    <p
      class="MuiTypography-root-81 MuiTypography-body1-90"
    >
      Click here for a new account
    </p>
  </a>
  <button
    class="MuiButtonBase-root-142 MuiButtonBase-disabled-143 MuiButton-root-116 MuiButton-contained-127 MuiButton-containedPrimary-128 MuiButton-raised-130 MuiButton-raisedPrimary-131 MuiButton-sizeSmall-139 MuiButton-disabled-136 {submit Loginform-submit-8"
    disabled=""
    name="Log in"
    tabindex="-1"
    type="submit"
  >
    <span
      class="MuiButton-label-117"
    >
      Go
    </span>
  </button>
</div>
`;
